File: core/ux/rendering/src/main/java/ru/fredboy/cavedroid/ux/rendering/GameRenderer.java
package ru.fredboy.cavedroid.ux.rendering;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.TimeUtils;
import org.jetbrains.annotations.Nullable;
import ru.fredboy.cavedroid.common.di.GameScope;
import ru.fredboy.cavedroid.common.utils.MeasureUnitsUtilsKt;
import ru.fredboy.cavedroid.common.utils.RenderingUtilsKt;
import ru.fredboy.cavedroid.domain.assets.usecase.GetFontUseCase;
import ru.fredboy.cavedroid.domain.configuration.model.CameraContext;
import ru.fredboy.cavedroid.domain.configuration.repository.ApplicationContextRepository;
import ru.fredboy.cavedroid.domain.configuration.repository.GameContextRepository;
import ru.fredboy.cavedroid.entity.mob.model.Player;
import ru.fredboy.cavedroid.game.controller.mob.MobController;
import ru.fredboy.cavedroid.game.window.TooltipManager;
import ru.fredboy.cavedroid.game.world.GameWorld;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;

@GameScope
public class GameRenderer {

    private static final float CAMERA_SPEED = 72f;
    private static final float MAX_CAM_DISTANCE_FROM_PLAYER = 64f;
    private final ApplicationContextRepository mApplicationContextRepository;
    private final GameContextRepository mGameContextRepository;
    private final MobController mMobsController;
    private final GameWorld mGameWorld;
    private final List<IGameRenderer> mRenderers;
    private final TooltipManager mTooltipManager;
    private final GetFontUseCase mGetFontUseCase;
    private final Vector2 mCamCenterToPlayer = new Vector2();
    private long mCameraDelayMs = 0L;

    private final ShapeRenderer shaper;
    private final SpriteBatch spriter;

    @Inject
    public GameRenderer(ApplicationContextRepository applicationContextRepository,
                GameContextRepository gameContextRepository,
                 MobController mobsController,
                 GameWorld gameWorld,
                 Set<IGameRenderer> renderers,
                 TooltipManager tooltipManager,
                 GetFontUseCase getFontUseCase) {
        @Nullable final CameraContext cameraContext = gameContextRepository.getCameraContext();

        shaper = new ShapeRenderer();
        spriter = new SpriteBatch();

        if (cameraContext != null) {
            shaper.setProjectionMatrix(cameraContext.getCamera().combined);
            spriter.setProjectionMatrix(cameraContext.getCamera().combined);
        } else {
            Gdx.app.error("GameRenderer", "Camera context was not set");
        }

        mApplicationContextRepository = applicationContextRepository;
        mGameContextRepository = gameContextRepository;
        mMobsController = mobsController;
        mGameWorld = gameWorld;
        mRenderers = new ArrayList<>(renderers);
        kotlin.collections.CollectionsKt.sortWith(mRenderers, new Comparator<IGameRenderer>() {
            @Override
            public int compare(IGameRenderer o1, IGameRenderer o2) {
                return o1.getRenderLayer() - o2.getRenderLayer();
            }
        });
        mTooltipManager = tooltipManager;
        mGetFontUseCase = getFontUseCase;

        Gdx.gl.glClearColor(0f, .6f, .6f, 1f);
    }

    private float getCamX() {
        @Nullable final CameraContext cameraContext = mGameContextRepository.getCameraContext();
        if (cameraContext != null) {
            return cameraContext.getViewport().x;
        } else {
            Gdx.app.error("GameRenderer", "Camera context was not set");
            return 0;
        }
    }

    private float getCamY() {
        @Nullable final CameraContext cameraContext = mGameContextRepository.getCameraContext();
        if (cameraContext != null) {
            return cameraContext.getViewport().y;
        } else {
            Gdx.app.error("GameRenderer", "Camera context was not set");
            return 0;
        }
    }

    private float getWidth() {
        return mApplicationContextRepository.getWidth();
    }
    private float getHeight() {
        return mApplicationContextRepository.getHeight();
    }

    private void setCamPos(float x, float y) {
        @Nullable final CameraContext cameraContext = mGameContextRepository.getCameraContext();

        if (cameraContext != null) {
            cameraContext.getCamera().position.set(x, y, 0);
            cameraContext.getViewport().x = x;
            cameraContext.getViewport().y = y;
        } else {
            Gdx.app.error("GameRenderer", "Camera context was not set");
        }
    }

    private Rectangle getCameraViewport() {
        @Nullable final CameraContext cameraContext = mGameContextRepository.getCameraContext();
        return cameraContext != null ? cameraContext.getViewport() : new Rectangle();
    }

    private void updateDynamicCameraPosition(float delta) {
        Player player = mMobsController.getPlayer();

        float plTargetX = player.getX() + player.getWidth() / 2;
        float plTargetY = player.getY() + player.getHeight() / 2;

        float camCenterX = getCamX() + getWidth() / 2;
        float camCenterY = getCamY() + getHeight() / 2;

        float camTargetX, camTargetY;

        boolean followPlayer = player.getControlMode() == Player.ControlMode.WALK || !mApplicationContextRepository.isTouch();

        if (followPlayer) {
            camTargetX = plTargetX + Math.min(player.getVelocity().x * 2, getWidth() / 2);
            camTargetY = plTargetY + player.getVelocity().y;
        } else {
            camTargetX = MeasureUnitsUtilsKt.getPx(player.getCursorX()) + MeasureUnitsUtilsKt.getPx(1) / 2;
            camTargetY = MeasureUnitsUtilsKt.getPx(player.getCursorY()) + MeasureUnitsUtilsKt.getPx(1) / 2;
        }

        Vector2 moveVector = new Vector2(camTargetX - camCenterX, camTargetY - camCenterY);

        if (followPlayer && player.getVelocity().isZero()) {
            mCameraDelayMs = TimeUtils.millis();
            mCamCenterToPlayer.x = plTargetX - camCenterX;
            mCamCenterToPlayer.y = plTargetY - camCenterY;
        }

        if (TimeUtils.timeSinceMillis(mCameraDelayMs) < 500L && !player.getVelocity().isZero()) {
            updateStaticCameraPosition(plTargetX - mCamCenterToPlayer.x,
                    camCenterY + moveVector.y * delta * 2);
            return;
        }

        float camX = getCamX();
        float camY = getCamY();
        float fullWorldPx = MeasureUnitsUtilsKt.getPx(mGameWorld.getWidth());
        float worldWidthScreenOffset = fullWorldPx - getWidth() / 2;

        if (moveVector.x >= worldWidthScreenOffset) {
            camX += fullWorldPx;
            moveVector.x -= fullWorldPx;
        } else if (moveVector.x <= -worldWidthScreenOffset) {
            camX -= fullWorldPx;
            moveVector.x += fullWorldPx;
        }

        setCamPos(camX + moveVector.x * delta * 2, camY + moveVector.y * delta * 2);


        camX = getCamX();
        camY = getCamY();

        if (camX + getWidth() / 2 > plTargetX + MAX_CAM_DISTANCE_FROM_PLAYER) {
            camX = plTargetX + MAX_CAM_DISTANCE_FROM_PLAYER - getWidth() / 2;
        }

        if (camY + getHeight() / 2 > plTargetY + MAX_CAM_DISTANCE_FROM_PLAYER) {
            camY = plTargetY + MAX_CAM_DISTANCE_FROM_PLAYER - getHeight() / 2;
        }

        if (camX + getWidth() / 2 < plTargetX - MAX_CAM_DISTANCE_FROM_PLAYER) {
            camX = plTargetX - MAX_CAM_DISTANCE_FROM_PLAYER - getWidth() / 2;
        }

        if (camY + getHeight() / 2 < plTargetY - MAX_CAM_DISTANCE_FROM_PLAYER) {
            camY = plTargetY - MAX_CAM_DISTANCE_FROM_PLAYER - getHeight() / 2;
        }

        setCamPos(camX, camY);
    }

    private void updateStaticCameraPosition(float targetX, float targetY) {
        setCamPos(targetX - getWidth() / 2, targetY - getHeight() / 2);
    }

    private void updateStaticCameraPosition() {
        Player player = mMobsController.getPlayer();

        updateStaticCameraPosition(player.getX() + player.getWidth() / 2,
                player.getY() + player.getHeight() / 2);
    }

    private void updateCameraPosition(float delta) {
        if (mApplicationContextRepository.useDynamicCamera()) {
            updateDynamicCameraPosition(delta);
        } else {
            updateStaticCameraPosition();
        }
    }

    private float transformScreenX(int screenX) {
        return getWidth() / Gdx.graphics.getWidth() * screenX;
    }

    private float transformScreenY(int screenY) {
        return getHeight() / Gdx.graphics.getHeight() * screenY;
    }

    private void handleMousePosition() {
        final float screenX = transformScreenX(Gdx.input.getX());
        final float screenY = transformScreenY(Gdx.input.getY());

        if (!mTooltipManager.getCurrentMouseTooltip().isEmpty()) {
            RenderingUtilsKt.drawString(spriter, mGetFontUseCase.invoke(),
                    mTooltipManager.getCurrentMouseTooltip(), screenX + 1, screenY + 1, Color.BLACK);
            RenderingUtilsKt.drawString(spriter, mGetFontUseCase.invoke(),
                    mTooltipManager.getCurrentMouseTooltip(), screenX, screenY, Color.WHITE);
        }
    }

    public void render(float delta) {
        updateCameraPosition(delta);

        if (mGameContextRepository.getJoystick() != null && mGameContextRepository.getJoystick().getActive()) {
            mGameContextRepository.getJoystick().updateState(
                    transformScreenX(Gdx.input.getX(mGameContextRepository.getJoystick().getPointer())),
                    transformScreenY(Gdx.input.getY(mGameContextRepository.getJoystick().getPointer()))
            );
        }

        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        spriter.begin();
        for (IGameRenderer iGameRenderer : mRenderers) {
            iGameRenderer.draw(spriter, shaper, getCameraViewport(), delta);
        }
        handleMousePosition();
        spriter.end();

    }

}

File: core/gameplay/rendering/src/main/kotlin/ru/fredboy/cavedroid/gameplay/rendering/GameRenderer.kt
package ru.fredboy.cavedroid.gameplay.rendering

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.physics.box2d.Box2DDebugRenderer
import com.badlogic.gdx.utils.Disposable
import ru.fredboy.cavedroid.common.di.GameScope
import ru.fredboy.cavedroid.common.utils.TooltipManager
import ru.fredboy.cavedroid.common.utils.drawString
import ru.fredboy.cavedroid.common.utils.ifTrue
import ru.fredboy.cavedroid.common.utils.meters
import ru.fredboy.cavedroid.domain.assets.usecase.GetFontUseCase
import ru.fredboy.cavedroid.domain.configuration.repository.ApplicationContextRepository
import ru.fredboy.cavedroid.domain.configuration.repository.GameContextRepository
import ru.fredboy.cavedroid.domain.world.model.Biome
import ru.fredboy.cavedroid.entity.mob.abstraction.PlayerAdapter
import ru.fredboy.cavedroid.game.world.GameWorld
import ru.fredboy.cavedroid.game.world.lighting.LightingSystem
import ru.fredboy.cavedroid.gameplay.rendering.renderer.hud.IHudRenderer
import ru.fredboy.cavedroid.gameplay.rendering.renderer.world.IWorldRenderer
import javax.inject.Inject
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.min

@GameScope
class GameRenderer @Inject constructor(
    private val applicationContextRepository: ApplicationContextRepository,
    private val gameContextRepository: GameContextRepository,
    private val tooltipManager: TooltipManager,
    private val gameWorld: GameWorld,
    private val lightingSystem: LightingSystem,
    private val player: PlayerAdapter,
    private val getFont: GetFontUseCase,
    @Suppress("LocalVariableName") _worldRenderers: Set<@JvmSuppressWildcards IWorldRenderer>,
    @Suppress("LocalVariableName") _hudRenderers: Set<@JvmSuppressWildcards IHudRenderer>,
) : Disposable {

    private val worldRenderers = _worldRenderers.sortedBy { it.renderLayer }
    private val hudRenderers = _hudRenderers.sortedBy { it.renderLayer }

    private val debugRenderer = ifTrue(applicationContextRepository.isDebug()) {
        Box2DDebugRenderer(
            /* drawBodies = */ true,
            /* drawJoints = */ false,
            /* drawAABBs = */ false,
            /* drawInactiveBodies = */ false,
            /* drawVelocities = */ false,
            /* drawContacts = */ true,
        )
    }

    private val camera = OrthographicCamera()
        .apply {
            setToOrtho(
                /* yDown = */ true,
                /* viewportWidth = */ applicationContextRepository.getWidth().meters,
                /* viewportHeight = */ applicationContextRepository.getHeight().meters,
            )
        }

    private val hudCamera = OrthographicCamera()
        .apply {
            setToOrtho(
                /* yDown = */ true,
                /* viewportWidth = */ applicationContextRepository.getWidth(),
                /* viewportHeight = */ applicationContextRepository.getHeight(),
            )
        }

    private val shaper = ShapeRenderer()

    private val spriter = SpriteBatch()

    private val baseSkyColor = Color()
    private val skyLeftColor = Color()
    private val skyRightColor = Color()

    init {
        Gdx.gl.glClearColor(0f, .6f, .6f, 1f)
    }

    private fun updateStaticCameraPosition(targetX: Float, targetY: Float) {
        camera.position.set(targetX, targetY, 0f)
    }

    private fun updateStaticCameraPositionToPlayer() {
        updateStaticCameraPosition(player.x, player.y)
    }

    private fun updateDynamicCameraPosition(delta: Float) {
        val cameraTargetPosition = Vector3().apply {
            x = player.x / 2f + player.aimX / 2f
            y = player.y / 2f + player.aimY / 2f
        }

        val moveVector = cameraTargetPosition.sub(camera.position)

        if (!moveVector.isZero(0.05f)) {
            val l = moveVector.len()
            moveVector.nor().scl(min(l, 30f * delta))
        }

        camera.position.add(moveVector)
    }

    private fun wrapCameraToPlayerSeam() {
        if (gameWorld.isInfinite) {
            // No seam to wrap across in an infinite world.
            return
        }

        val worldWidth = gameWorld.width.toFloat()
        val cameraToPlayer = player.x - camera.position.x
        if (cameraToPlayer > worldWidth / 2f) {
            camera.position.add(Vector3(worldWidth, 0f, 0f))
        } else if (cameraToPlayer < -worldWidth / 2f) {
            camera.position.sub(Vector3(worldWidth, 0f, 0f))
        }
    }

    private fun updateCameraPosition(delta: Float) {
        wrapCameraToPlayerSeam()
        if (applicationContextRepository.useDynamicCamera()) {
            updateDynamicCameraPosition(delta)
        } else {
            updateStaticCameraPositionToPlayer()
        }
        gameContextRepository.getCameraContext().viewport.apply {
            x = hudCamera.position.x - hudCamera.viewportWidth / 2
            y = hudCamera.position.y - hudCamera.viewportHeight / 2
            width = hudCamera.viewportWidth
            height = hudCamera.viewportHeight
        }
        gameContextRepository.getCameraContext().visibleWorld.set(getVisibleWorldRect())
        camera.update()
    }

    private fun getVisibleWorldRect(): Rectangle {
        return Rectangle(
            camera.position.x - camera.viewportWidth / 2,
            camera.position.y - camera.viewportHeight / 2,
            camera.viewportWidth,
            camera.viewportHeight,
        )
    }

    private fun handleMousePosition() {
        val viewportX = hudCamera.viewportWidth / Gdx.graphics.width * Gdx.input.x
        val viewportY = hudCamera.viewportHeight / Gdx.graphics.height * Gdx.input.y

        if (tooltipManager.currentMouseTooltip.isNotEmpty()) {
            spriter.drawString(
                font = getFont(),
                str = tooltipManager.currentMouseTooltip,
                x = viewportX + 1f,
                y = viewportY + 1f,
                color = Color.BLACK,
            )
            spriter.drawString(
                font = getFont(),
                str = tooltipManager.currentMouseTooltip,
                x = viewportX,
                y = viewportY,
                color = Color.WHITE,
            )
        }
    }

    fun onResize() {
        camera.setToOrtho(
            /* yDown = */ true,
            /* viewportWidth = */ applicationContextRepository.getWidth().meters,
            /* viewportHeight = */ applicationContextRepository.getHeight().meters,
        )
        hudCamera.setToOrtho(
            /* yDown = */ true,
            /* viewportWidth = */ applicationContextRepository.getWidth(),
            /* viewportHeight = */ applicationContextRepository.getHeight(),
        )
        resetCameraToPlayer()
    }

    private fun updateSkyColors() {
        val sunlight = gameWorld.getSunlight()
        val normalizedTime = gameWorld.getNormalizedTime()

        val base = baseSkyColor.apply {
            if (sunlight < 0.5f) {
                set(MIDNIGHT_COLOR).lerp(HORIZON_COLOR, sunlight * 2f)
            } else {
                set(HORIZON_COLOR).lerp(NOON_COLOR, (sunlight - 0.5f) * 2f)
            }
        }

        val twilight = (1f - 2f * abs(sunlight - 0.5f)).coerceIn(0f, 1f)
        val warmSide = -cos(normalizedTime * PI.toFloat())
        val warmBlend = twilight * abs(warmSide)
        val darkBlend = warmBlend * 0.5f

        if (warmSide < 0f) {
            skyLeftColor.set(base).lerp(HORIZON_COLOR, warmBlend)
            skyRightColor.set(base).lerp(MIDNIGHT_COLOR, darkBlend)
        } else {
            skyLeftColor.set(base).lerp(MIDNIGHT_COLOR, darkBlend)
            skyRightColor.set(base).lerp(HORIZON_COLOR, warmBlend)
        }

        val biomeFactor = gameWorld.biomeProximityFactor(
            centerX = player.x,
            rangeBlocks = WEATHER_SKY_FADE_BLOCKS,
        ) { it != Biome.DESERT }
        val intensity = gameWorld.weatherIntensity * biomeFactor
        if (intensity > 0f) {
            val factor = 1f - intensity * (1f - RAIN_SKY_DARKENING)
            skyLeftColor.r *= factor
            skyLeftColor.g *= factor
            skyLeftColor.b *= factor
            skyRightColor.r *= factor
            skyRightColor.g *= factor
            skyRightColor.b *= factor
        }
    }

    private fun drawSky() {
        val x = camera.position.x - camera.viewportWidth / 2f
        val y = camera.position.y - camera.viewportHeight / 2f
        shaper.begin(ShapeRenderer.ShapeType.Filled)
        shaper.rect(
            /* x = */ x,
            /* y = */ y,
            /* width = */ camera.viewportWidth,
            /* height = */ camera.viewportHeight,
            /* col1 = */ skyLeftColor,
            /* col2 = */ skyRightColor,
            /* col3 = */ skyRightColor,
            /* col4 = */ skyLeftColor,
        )
        shaper.end()
    }

    fun render(delta: Float) {
        updateCameraPosition(delta)

        spriter.projectionMatrix = camera.combined
        shaper.projectionMatrix = camera.combined

        Gdx.gl.glClearColor(0f, 0f, 0f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)

        updateSkyColors()
        drawSky()

        spriter.begin()
        val cameraViewport = Rectangle(
            /* x = */ camera.position.x - camera.viewportWidth / 2,
            /* y = */ camera.position.y - camera.viewportHeight / 2,
            /* width = */ camera.viewportWidth,
            /* height = */ camera.viewportHeight,
        )
        worldRenderers.forEach { renderer -> renderer.draw(spriter, shaper, cameraViewport, delta) }

        spriter.end()

        lightingSystem.render(camera)

        if (gameContextRepository.shouldShowInfo()) {
            debugRenderer?.render(gameWorld.world, camera.combined)
        }

        spriter.projectionMatrix = hudCamera.combined
        shaper.projectionMatrix = hudCamera.combined

        spriter.begin()
        val hudCameraViewport = Rectangle(
            /* x = */ hudCamera.position.x - hudCamera.viewportWidth / 2,
            /* y = */ hudCamera.position.y - hudCamera.viewportHeight / 2,
            /* width = */ hudCamera.viewportWidth,
            /* height = */ hudCamera.viewportHeight,
        )
        hudRenderers.forEach { renderer -> renderer.draw(spriter, shaper, hudCameraViewport, delta) }
        handleMousePosition()
        spriter.end()
    }

    fun resetCameraToPlayer() {
        updateStaticCameraPositionToPlayer()
        camera.update()
    }

    override fun dispose() {
        spriter.dispose()
        shaper.dispose()
        debugRenderer?.dispose()
        worldRenderers.forEach { renderer -> renderer.dispose() }
        hudRenderers.forEach { renderer -> renderer.dispose() }
    }

    companion object {
        private val MIDNIGHT_COLOR = Color(0f, 0f, 0.1f, 1f)
        private val HORIZON_COLOR = Color(0.9f, 0.45f, 0.25f, 1f)
        private val NOON_COLOR = Color(0.4f, 0.7f, 1f, 1f)

        private const val RAIN_SKY_DARKENING = 0.3f
        private const val WEATHER_SKY_FADE_BLOCKS = 16f

        private const val CAMERA_DELTA = 1f / 60f
    }
}
